features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,318 @@
+- Compatible to old API interface
+
+# Overview
+
+Container contains a local store for caching resources. The local store caches resources by resource id.
+
+### Store Update
+
+When SDK recieves result from server, for each record, the newly fetched record would always replace the old one with the same record id if there is one.
+
+### API
+
+For fetch API of each resource, there would be a new callback that should return result from cache immediately.
+
+The interface for each language may be different.
+
+```
+getResource(params, callback, cachedCallback?)
+```
+
+Resource includes conversation and message.
+
+The cachedCallback should be optional, thus compatible to the old API interace.
+
+After client app receives result from both callback, client app need to resolve a single result for display. Skygear Chat Container would provide some utility functions to help client app resolve two callback result.
+
+### Expected UI Display / Common Use Case
+
+Client App is expected to keep a single list of messages for display in table. And they are required to keep track to the newest and oldest fetched messages to determine when to fetch result from server.
+
+Client App should replace old message with new message if the same message id is found.
+
+Deleted messages are also returned in fetch API, with empty content and marked as deleted. This is to keep the whole message list complete. Client App is expected to skip these message in UI display.
+
+We may provide a high level API **ManagedMessageList** to handle this for the developer.
+
+# Sample Code for Message List, with ManagedMessageList
+
+*All below UI update, I use `this.tableView.reloadTable()`. It is up to client app to decide how to update the UI*
+
+#### Enter conversation
+
+```
+this.stickToBottom = true | false
+this.messageList = new ManagedMessageList(container: skygear.defaultContainer(),
+                                          conversation: this.conversation,
+                                          delegate: this)
+
+this.messageList.fetchLatestMessages()
+this.messageList.startSubscribing()
+
+// implement ManagedMessageListDelegate
+func managedMessageListDidChange(messageDeltas: [MessageDelta]) {
+  this.tableView.reloadTable()
+  if (this.stickToBottom) {
+    this.tableView.scrollToBottom()
+  }
+}
+```
+
+#### Render table
+
+```
+func numberOfCells() -> Int {
+  return this.messageList.messageCount()
+}
+
+func cellForRow(index: Int) -> Cell {
+  let message = this.messageList.messageAt(index)
+
+  let cell = /* render cell with message */
+
+  return cell
+}
+```
+
+#### Leave conversation
+
+```
+this.messageList.stopSubscribing()
+```
+
+# Sample Code for Conversation List
+
+### In conversation list
+
+#### Start subscribing conversation changes
+
+When
+- enter conversation list
+
+```
+skygearChat.subscribeConversations(completion: func (conversationDeltas) {
+  this.conversations = applyDelta(conversationDeltas, toList: this.conversations)
```

Comment:
What is `applyDelta`? Is it easier to just get all conversations managed like message list?